File: c/physics/bounce/test/test_physics.c
/**
 * @file   test_physics.c
 * @brief  Unit tests for bounce physics (ball_update, ball_speed).
 */

#include "unity.h"
#include "../physics.h"
#include <math.h>

void setUp(void)    {}
void tearDown(void) {}

void test_normal_movement(void) {
    Ball b = {100.0f, 100.0f, 4.0f, 3.0f};
    ball_update(&b, 10, 800, 600);
    TEST_ASSERT_FLOAT_WITHIN(0.01f, 104.0f, b.x);
    TEST_ASSERT_FLOAT_WITHIN(0.01f, 103.0f, b.y);
}

void test_left_wall_reflection(void) {
    Ball b = {10.0f, 100.0f, -5.0f, 3.0f};
    ball_update(&b, 10, 800, 600);
    TEST_ASSERT_FLOAT_WITHIN(0.01f, 5.0f, b.vx);
}

void test_right_wall_reflection(void) {
    Ball b = {794.0f, 100.0f, 5.0f, 3.0f};
    ball_update(&b, 10, 800, 600);
    TEST_ASSERT_FLOAT_WITHIN(0.01f, -5.0f, b.vx);
}

void test_top_wall_reflection(void) {
    Ball b = {100.0f, 8.0f, 4.0f, -5.0f};
    ball_update(&b, 10, 800, 600);
    TEST_ASSERT_FLOAT_WITHIN(0.01f, 5.0f, b.vy);
}

void test_bottom_wall_reflection(void) {
    Ball b = {100.0f, 594.0f, 4.0f, 5.0f};
    ball_update(&b, 10, 800, 600);
    TEST_ASSERT_FLOAT_WITHIN(0.01f, -5.0f, b.vy);
}

void test_speed_is_pythagorean(void) {
    Ball b = {0, 0, 3.0f, 4.0f};
    TEST_ASSERT_FLOAT_WITHIN(0.001f, 5.0f, ball_speed(&b));
}

void test_speed_does_not_change_on_bounce(void) {
    Ball b = {10.0f, 100.0f, -3.0f, 4.0f};
    float speed_before = ball_speed(&b);
    ball_update(&b, 10, 800, 600);
    TEST_ASSERT_FLOAT_WITHIN(0.001f, speed_before, ball_speed(&b));
}

int main(void) {
    UNITY_BEGIN();
    RUN_TEST(test_normal_movement);
    RUN_TEST(test_left_wall_reflection);
    RUN_TEST(test_right_wall_reflection);
    RUN_TEST(test_top_wall_reflection);
    RUN_TEST(test_bottom_wall_reflection);
    RUN_TEST(test_speed_is_pythagorean);
    RUN_TEST(test_speed_does_not_change_on_bounce);
    return UNITY_END();
}

File: python/physics/bounce/tests/test_physics.py
"""Unit tests for bounce physics (update, speed)."""

import math
import pytest
from physics import Ball, update, speed


def test_normal_movement() -> None:
    b = Ball(x=100.0, y=100.0, vx=4.0, vy=3.0)
    update(b, radius=10, width=800, height=600)
    assert math.isclose(b.x, 104.0, abs_tol=0.01)
    assert math.isclose(b.y, 103.0, abs_tol=0.01)


def test_left_wall_reflection() -> None:
    b = Ball(x=10.0, y=100.0, vx=-5.0, vy=3.0)
    update(b, radius=10, width=800, height=600)
    assert math.isclose(b.vx, 5.0, abs_tol=0.01), f"expected vx=5.0, got {b.vx}"


def test_right_wall_reflection() -> None:
    b = Ball(x=794.0, y=100.0, vx=5.0, vy=3.0)
    update(b, radius=10, width=800, height=600)
    assert math.isclose(b.vx, -5.0, abs_tol=0.01), f"expected vx=-5.0, got {b.vx}"


def test_top_wall_reflection() -> None:
    b = Ball(x=100.0, y=8.0, vx=4.0, vy=-5.0)
    update(b, radius=10, width=800, height=600)
    assert math.isclose(b.vy, 5.0, abs_tol=0.01), f"expected vy=5.0, got {b.vy}"


def test_bottom_wall_reflection() -> None:
    b = Ball(x=100.0, y=594.0, vx=4.0, vy=5.0)
    update(b, radius=10, width=800, height=600)
    assert math.isclose(b.vy, -5.0, abs_tol=0.01), f"expected vy=-5.0, got {b.vy}"


def test_speed_is_pythagorean() -> None:
    b = Ball(x=0.0, y=0.0, vx=3.0, vy=4.0)
    assert math.isclose(speed(b), 5.0, abs_tol=0.001)


def test_speed_preserved_on_bounce() -> None:
    b = Ball(x=10.0, y=100.0, vx=-3.0, vy=4.0)
    before = speed(b)
    update(b, radius=10, width=800, height=600)
    assert math.isclose(speed(b), before, abs_tol=0.001), "speed changed on elastic bounce"
